core/deployment/src/test/java/io/quarkiverse/langchain4j/test/StreamingChatLanguageModelToolTest.java

```diff
@@ -0,0 +1,115 @@
+package io.quarkiverse.langchain4j.test;
+
+import static io.restassured.RestAssured.get;
+import static org.hamcrest.Matchers.containsString;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.function.Supplier;
+
+import jakarta.inject.Singleton;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import dev.langchain4j.agent.tool.Tool;
+import dev.langchain4j.agent.tool.ToolExecutionRequest;
+import dev.langchain4j.agent.tool.ToolSpecification;
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.data.message.ChatMessage;
+import dev.langchain4j.data.message.ToolExecutionResultMessage;
+import dev.langchain4j.data.message.UserMessage;
+import dev.langchain4j.model.StreamingResponseHandler;
+import dev.langchain4j.model.chat.StreamingChatLanguageModel;
+import dev.langchain4j.model.output.Response;
+import io.quarkiverse.langchain4j.RegisterAiService;
+import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+
+public class StreamingChatLanguageModelToolTest {
+    private static final String TOOL_NAME = "tool";
+
+    @Path("/test")
+    static class MyResource {
+
+        private final MyService service;
+
+        MyResource(MyService service) {
+            this.service = service;
+        }
+
+        @GET
+        public Multi<String> streaming() {
+            return service.chat("what is the Answer to the Ultimate Question of Life, the Universe, and Everything?");
```

Comment:
I would be interested in seeing which thread the items are emitted on when tools are used.

Reply:
Currently, all calls are executed on the event-loop thread. 

So, waiting in a call to generate a response will also fail: 

```
    public static class MyModelSupplier implements Supplier<StreamingChatLanguageModel> {
        @Override
        public StreamingChatLanguageModel get() {
            return (messages, handler) -> {
                handler.onNext("4");

                // block thread intentionally to verify there is no: The current thread cannot be blocked: vert.x-eventloop-thread-1
                Uni.createFrom().voidItem().onItem().delayIt().by(Duration.ofMillis(500)).await().indefinitely();

                handler.onNext("2");
                handler.onComplete(new Response<>(new AiMessage("")));
            };
        }
    }
```

However, this is probably not something we need to support, as the API is callback-based.

A nicer implementation would be 

```
            return (messages, handler) -> {
                Multi.createFrom().items("4", "2").onCompletion().invoke(() -> {
                            handler.onComplete(new Response<>(new AiMessage("")));
                        })
                        .subscribe().with(handler::onNext);
```

This is different for tool calls, though, where a response needs to be returned blocking, as all calls after `handler.onComplete(Response.from(AiMessage.from(toolExecutionRequest)));` run synchronously on one thread. 

The question is, what is the benefit of running on the event loop?

Here are some logs, executing the test case from the main branch:

```
2024-09-02 15:19:20,973 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (vert.x-eventloop-thread-1) requesting tool execution
2024-09-02 15:19:20,976 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyTool] (vert.x-eventloop-thread-1) executing tool call
2024-09-02 15:19:20,983 ERROR [io.qua.lan.run.too.QuarkusToolExecutor] (vert.x-eventloop-thread-1) Error while executing tool 'class io.quarkiverse.langchain4j.test.StreamingChatLanguageModelToolTest$MyTool': java.lang.IllegalStateException: The current thread cannot be blocked: vert.x-eventloop-thread-1
```

And here, after the change to `Infrastructure.getDefaultWorkerPool()`:

```
2024-09-02 15:21:36,177 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (executor-thread-1) requesting tool execution
2024-09-02 15:21:36,180 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyTool] (executor-thread-1) executing tool call
2024-09-02 15:21:37,124 INFO  [io.qua.lan.tes.StreamingChatLanguageModelToolTest$MyModelSupplier$1] (executor-thread-1) processing tool execution result
```

Note that there is already some offloading to `Infrastructure.getDefaultWorkerPool()` in [this line of code](https://github.com/quarkiverse/quarkus-langchain4j/blob/2b014289dd3f8c2d14bc6494bae76a0b66a0cbea/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java#L151).

After looking again through the code, I see the options of  
 * running `AiServiceTokenStream` on `Infrastructure.getDefaultWorkerPool` as proposed here. 
 * requesting an API extension from langchain to support an asynchronous API for tool executions.

Reply:
I missed your comment https://github.com/quarkiverse/quarkus-langchain4j/pull/846#issuecomment-2324667832, let me check....

Reply:
> Note that there is already some offloading to Infrastructure.getDefaultWorkerPool() in [this line of code](https://github.com/quarkiverse/quarkus-langchain4j/blob/2b014289dd3f8c2d14bc6494bae76a0b66a0cbea/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceMethodImplementationSupport.java#L151).

Indeed, but that happens before the `Multi` is created. 

> requesting an API extension from langchain to support an asynchronous API for tool executions.

Yeah, this is a far better option as `running AiServiceTokenStream on Infrastructure.getDefaultWorkerPool as proposed here` is not the proper way to go about this

Reply:
Resolving this conversation in favor of the main conversation on the PR.